File: rules/llm06-excessive-agency/typescript/dangerous-sink-in-tool.ts
import { tool } from "ai";
import { exec, execSync } from "child_process";
import { z } from "zod";

export const shellTool = tool({
  description: "Run a shell command",
  inputSchema: z.object({ command: z.string() }),
  execute: async ({ command }) => {
    // ruleid: dangerous-sink-in-tool-js
    return execSync(command).toString();
  },
});

export const evalTool = tool({
  description: "Evaluate a JavaScript expression",
  inputSchema: z.object({ expression: z.string() }),
  execute: ({ expression }) => {
    // ruleid: dangerous-sink-in-tool-js
    return eval(expression);
  },
});

export const wordCountTool = tool({
  description: "Count words",
  inputSchema: z.object({ text: z.string() }),
  execute: async ({ text }) => {
    // ok: dangerous-sink-in-tool-js
    return text.split(/\s+/).length;
  },
});

export const fixedCommandTool = tool({
  description: "List the working directory",
  inputSchema: z.object({}),
  execute: async () => {
    // ok: dangerous-sink-in-tool-js
    return execSync("ls -la").toString();
  },
});

File: rules/llm06-excessive-agency/python/dangerous-sink-in-tool.py
import os
import subprocess

from langchain_core.tools import tool
from mcp.server.fastmcp import FastMCP

mcp = FastMCP("demo")


@tool
def run_python(code: str) -> str:
    """Execute Python code and return the result."""
    # ruleid: dangerous-sink-in-tool
    return str(eval(code))


@tool("shell", return_direct=True)
def run_shell(command: str) -> str:
    """Run a shell command."""
    # ruleid: dangerous-sink-in-tool
    return subprocess.check_output(command, shell=True).decode()


@mcp.tool()
def system_command(cmd: str) -> int:
    """Run a system command."""
    # ruleid: dangerous-sink-in-tool
    return os.system(cmd)


@mcp.tool(description="execute a snippet")
def exec_snippet(snippet: str) -> None:
    # ruleid: dangerous-sink-in-tool
    exec(snippet)


# CrewAI exports the same @tool decorator shape (from crewai.tools import tool):
@tool("CrewAI shell tool")
def crewai_shell(command: str) -> str:
    """Run a command for the crew."""
    # ruleid: dangerous-sink-in-tool
    return subprocess.run(command, shell=True, capture_output=True).stdout.decode()


@tool
def word_count(text: str) -> int:
    """Count words in text."""
    # ok: dangerous-sink-in-tool
    return len(text.split())


@tool
def list_directory(path: str) -> str:
    """List a directory using a fixed command with an args list."""
    # ok: dangerous-sink-in-tool
    return subprocess.check_output(["ls", "-la"]).decode()


def not_a_tool(cmd: str) -> int:
    # ok: dangerous-sink-in-tool
    return os.system(cmd)
